Temporary passwords for public non-anonymous protected shares (ie: files shared with an email recipient).

StCyr opened the pull request:
This PR implements https://github.com/nextcloud/server/issues/31005

With this change, passwords protecting public non-anonymous shares become temporary: They have an expiration time and a background job changes them to some random undisclosed value when their expiration time is passed.

To request a new temporary password, recipients use a new "request password" button that is added to protected public non-anonymous shares' authentication page. This button serves the same functionality as the "request password" button implemented by Talk (ie: To identify the person requesting the password), except that, here, the person identifies herself by proving she knows the email address with which the file is shared with (ie: she types her email address, and if it matches the email address with which the file is shared the person is considered "identified" and a new temporary password is sent to that email address). Of course, identification via a Talk session takes precedence when the share has its "video verification" attribute checked.

I'm waiting for your reviews and welcome your comments which I hope will be constructive 

Cyrille

===============================
UPDATE 20220217:

1. The existing protected public non-anonymous shares are not impacted by this change (they don't have a password_expiration_time, so the expiration background job doesn't consider them)
2. The temporary passwords are valid for 1 day.

===============================
UPDATE 20220223:
I've created a new branch: https://github.com/nextcloud/server/tree/enhancement/31005/temporary-passwords-v2

Compared to this branch, this new branch brings the following:

1. password_expiration_time is now checked during password check. This also makes the expiration background job not necessary anymore;
2. Temporary passwords validity period is now configurable via a system value

StCyr committed:
Adds a "Request password" button to the public share authentication page for shares
of type TYPE_EMAIL, when the "video verification" checkbox isn't checked. Users accessing
non-anonymous public shares (TYPE_EMAIL shares) can now request a temporary password themselves.

- Creates a migration step for the files_sharing app to add the 'password_expiration_time'
  attribute to the oc_shares table.
- Makes share temporary passwords' expiration time configurable via a system value.
- Adds a system config value to allow permanent share passwords

-Fixes a typo in a comment in apps/files_sharing/src/components/SharingEntryLink.vue

See https://github.com/nextcloud/server/issues/31005

Signed-off-by: Cyrille Bollu <cyrpub@bollu.be>

nickvergessen commented on apps/files_sharing/lib/BackgroundJob/ResetExpiredPasswordsJob.php:
```diff
@@ -0,0 +1,72 @@
+
+		$result = $qb->execute();
+		while ($row = $result->fetch()) {
+			// QUESTION: SHALL I RESPECT PASSWORD POLICY HERE (WHEN USED)?
```
I think you have to

StCyr replied:
It seems much more easier than I thought 😅

PVince81 commented:
did you forget to push something ?
<img width="554" alt="image" src="https://user-images.githubusercontent.com/277525/156143513-9463e939-61ce-43fe-867b-4dec2d8550ca.png">

- [ ] please run `build/autoloaderchecker.sh`

PVince81 commented on apps/files_sharing/lib/BackgroundJob/ResetExpiredPasswordsJob.php:
```diff
@@ -0,0 +1,86 @@
+use \OCP\Security\IHasher;
+use \OCP\Security\ISecureRandom;
+
+class ResetExpiredPasswordsJob extends TimedJob {
```
you can ignore these warnings, we'll need to fix the base class

StCyr replied:
Hmm I've a general remark about the necessity of this background job: 

I've created another branch (https://github.com/nextcloud/server/commits/enhancement/31005/temporary-passwords-v2), building on this one and adding 2 commits, where password's expiration time is checked in IManager::checkPassword (see https://github.com/nextcloud/server/commit/576de4bb8098777b4725921604be259add0eb95a) hence making this background job unnecessary anymore.

I believe it's a better solution than this background job. Could you please have a look and take a stand?

PVince81 replied:
@StCyr yes, let's use your approach without the background job which checks on-demand

StCyr replied:
on-demand check implemented (via 71ae5a077cb788a16d5f90ba2d958eb208900e5f and 9ef1c59c61249d1818087526f0d3b01ab4790196), together with configurable password expiration time

PVince81 commented on apps/files_sharing/lib/BackgroundJob/ResetExpiredPasswordsJob.php:
```diff
@@ -0,0 +1,86 @@
+		$qb = $this->connection->getQueryBuilder();
+
+		// QUESTION: DOES THE DATETIME COMPARAISON WORK WELL WHEN TIMEZONES ENTER THE GAME?
+		// I THINK SO, BECAUSE EVERYTHING HAPPENS ON THE SERVER, HENCE ON THE SAME TZ
```
usually the timezone of server is supposed to be UTC, so should be fine

PVince81 replied:
please remove the comment from the code when ready :smile:

StCyr replied:
Well, the background job idea is discarded => https://github.com/nextcloud/server/pull/31220#discussion_r817814581

PVince81 commented on apps/files_sharing/lib/Controller/ShareController.php:
```diff
@@ -209,6 +209,40 @@ protected function showAuthFailed(): TemplateResponse {
+	protected function generatePassword() {
+		// Generates a password respecting any password policy defined
+		$event = new \OCP\Security\Events\GenerateSecurePasswordEvent();
+	$this->eventDispatcher->dispatchTyped($event);
```
please fix indent

StCyr replied:
fixed in 2bc5788ef0599c5290f0e27557bfa63cd1372cee

PVince81 commented on apps/files_sharing/lib/Controller/ShareController.php:
```diff
@@ -209,6 +209,40 @@ protected function showAuthFailed(): TemplateResponse {
+		// Generates a password respecting any password policy defined
+		$event = new \OCP\Security\Events\GenerateSecurePasswordEvent();
+	$this->eventDispatcher->dispatchTyped($event);
+		$password = $event->getPassword() ?? $password = \OC::$server->getSecureRandom()->generate(20);
```
there's a second "=" in the line, please double check

StCyr replied:
fixed in 2bc5788ef0599c5290f0e27557bfa63cd1372cee

come-nc commented on apps/files_sharing/lib/Controller/ShareController.php:
```diff
@@ -209,6 +209,40 @@ protected function showAuthFailed(): TemplateResponse {
+		// Generates a password respecting any password policy defined
+		$event = new \OCP\Security\Events\GenerateSecurePasswordEvent();
+	$this->eventDispatcher->dispatchTyped($event);
+		$password = $event->getPassword() ?? $password = \OC::$server->getSecureRandom()->generate(20);
```
```suggestion
		$password = $event->getPassword() ?? \OC::$server->getSecureRandom()->generate(20);
```

nickvergessen replied:
and inject in constructor instead of using `\OC::$server`

StCyr replied:
fixed in 2bc5788ef0599c5290f0e27557bfa63cd1372cee

PVince81 commented:
- [ ] BUG: enter key not working when entering email address to request from

PVince81 commented:
- [ ] BUG: need to clear the (public) session info as soon as opening the "request password" page

Steps:

1. Share a folder with user "another@example.com"
2. Set a password "test"
3. Open the email
4. Open the link
5. Enter the password "test" => this authenticates you for the link
6. Open the link again
7. Click "request access"
8. Enter email address

Expected result: receive email with new password
Actual result: no email and the share folder appears directly, because of the previous session data

PVince81 commented on apps/files_sharing/lib/Controller/ShareController.php:
```diff
@@ -209,6 +209,40 @@ protected function showAuthFailed(): TemplateResponse {
+		return $this->shareManager->validateIdentity($this->share, $identityToken);
+	}
+
+	protected function generatePassword() {
```
looks like this method is unused ?

StCyr replied:
It's declared abstract in AuthPublicShareController.php and used there.

PVince81 replied:
alright, I didn't realize that ShareController was already extending it as I didn't see it in the diff. all good

PVince81 commented on lib/public/Share/IManager.php:
```diff
@@ -60,11 +60,14 @@ public function createShare(IShare $share);
 	 * @since 9.0.0
 	 */
-	public function updateShare(IShare $share);
+	public function updateShare(IShare $share, bool $sendPassword = false);
```
this will likely break implementations and might need a new version of the interface

wondering if there's another way by having or deriving this info from the IShare object directly

PVince81 replied:
before adding a new interface let's check if there are really more implementations of IManager or just one

PVince81 replied:
might be ok, see https://github.com/nextcloud/server/pull/31447

StCyr replied:
Hmmm.... Anyway, I'm wondering if this new feature (ie: not sending a temporary password upon public authenticate share creation) is really needed... Eventually, it's not that much of an issue if a temporary password is sent to the recipient upon share creation, rather than waiting for the recipient to explicitely requesting it.

I'm thinking about reverting this change. What's your opinion?

PVince81 replied:
there are scenarios where one might want to manually set a specific password to make it easier for the other person to type (or it's something they gave them already on paper or another medium), so we should allow for still supporting such scenarios

now, I'm not sure if this applies to this piece of code specifically ?

StCyr replied:
That's my point: I initially thought it was necessary to drop support for this scenario. So, we are now on the same line and commit ea4531aaaa6eb9fb3859e05b69ab773bfbfe7437 restores that functionality

PVince81 replied:
perfect, thanks a lot!

StCyr replied:
Hi @PVince81,

I'm unresolving this discussion because we would like to implement this feature of not sending the password via email in some cases eventually, and we understand that modifying the IManager interface is not an issue.

Also, we would like to make it clear that this PR, as it is currently implemented, completly drops support for permanent passwords for shares shared via email (except for shares created before the implementation of this PR): Once this PR gets implemented, all passwords set to protect a share shared via email will be temporary and expire after the delay specified by the `share_temporary_password_expiration_interval` system value, or 15 minutes by default; Your comment  https://github.com/nextcloud/server/pull/31220#discussion_r820812356 (specifically the phrase `(or it's something they gave them already on paper or another medium)`) makes us think we might not be aligned on this point.

PVince81 replied:
hi Cyrille, you're right about misalignment, thanks for spotting

I'll get this clarified internally as this would be a change of behavior

skjnldsv replied:
Hey @StCyr, after a bit of thoughts, we decided we would need a config switch.
Old behaviour would be off by default. :)

StCyr replied:
> Hey @StCyr, after a bit of thoughts, we decided we would need a config switch. Old behaviour would be off by default. :)

A new checkbox in the share menu?

![image](https://user-images.githubusercontent.com/8179031/159022812-8db57106-dcb1-40ed-a3ba-96e2354d5077.png)

PVince81 replied:
rather a config switch in config.php to switch it globally

StCyr replied:
implemented in c9940b194af8c59e0355c1d2660e614c67d99af6

I believe the PR is feature complete now. Do you want me to rebase and squash it?

PVince81 commented on lib/public/Share/IManager.php:
```diff
@@ -202,6 +205,15 @@ public function getShareById($id, $recipient = null);
+	 * @param string $recipient
+	 * @return bool
+	 */
+	public function validateIdentity(IShare $share, $identityToken);
```
this will need a new version of the interface to add new methods

I know we have multiple implementations of share provider, but not sure if we have multiple share managers or just one

PVince81 reviewed (changes requested):
I've added some comments and found some issues, please have a look

StCyr commented on core/js/publicshareauth.js:
```diff
@@ -1,3 +1,27 @@
+	emailPrompt.style.display="block";
+
+	// Hides password prompt
+	var passwordRequestButton = document.getElementById('request-password-button-not-talk');
```
if you want. But, I believe it's best to keep it technology-agnostic for if later we implement other means to request passwords (eg: by sms)

PVince81 commented on apps/files_sharing/lib/Controller/ShareController.php:
```diff
@@ -209,6 +215,62 @@ protected function showAuthFailed(): TemplateResponse {
+	 * @param string $identityToken
+	 * @return bool
+	 */
+	protected function validateIdentity(string $identityToken): bool {
```
please remove the type hint to allow null

StCyr replied:
you mean the `@return bool`?

PVince81 commented:
it might be ok to extend the IManager interface as it's only "for use", see new discussion here: https://github.com/nextcloud/server/pull/31447

StCyr commented:
> @StCyr when I try setting up Nextcloud I get this:
> 
> ```
> PHP Fatal error:  Declaration of OCA\FederatedFileSharing\FederatedShareProvider::create(OCP\Share\IShare $share) must be compatible with OCP\Share\IShareProvider::create(OCP\Share\IShare $share, bool $sendPassword = true) in /srv/www/htdocs/server/apps/federatedfilesharing/lib/FederatedShareProvider.php on line 171
> Installation failed
> ```
> 
> can you have a look ?
> 
> once this is resolved I think you can squash

Fixed in afbda38543e99e03d931dbb62cdcf07d9f3d7b48 

But I've seen that spreed's RoomShareProvider also implements the IShareProvider interface... Maybe I should avoid modifying this interface eventually...

StCyr commented:
Hello @PVince81,

Shall I rebase and squash now?

Do you want me to write a summary of the functional changes that have been implemented? Something else?

StCyr commented:
Here's a summary of the functional changes implemented:

- files_sharing createShare API:
  make password temporary for email shares (unless otherwise instructed by system config)
- files_sharing updateShare API:
  make password temporary for email shares (unless otherwise instructed by system config)
- sharebymail create: 
	Only send password when permanent password is enforced via system config; When using temporary passwords, 
	the password shall only be sent to the guest when she requests it from the UI.
	Note that sharebymail's update() function isn't changed, allowing the share owner to set an arbitrary temporary
	password and have it sent to the guest anyway.
- Manager checkPassword:
	When verifying the password of a public share, make sure it hasn't expired first.
- showShare API:
	Changes to allow guest to ask a temporary password. To do so, guest will do the following:
		1- Click on a new "Request password" button
		2- Upon clicking on the button, the guest is prompted to enter his email address
			2a- If the entered email address matches the share's shared_with property, a new temporary
			    share password is generated and sent to the user via email.
			2b- If the entered email address doesn't match the share's shared_with property, the guest
			    is presented with an error message telling him that he's not authorized to request a
			    password for this share.
		3- At any time, the user can click on a new "Back" button to return to the password prompt page.

PVince81 commented:
@StCyr CI is failing, can you squash and rebase ?

PVince81 commented:
never mind, I see you did one hour ago. weird

PVince81 commented:
I saw a bunch of those, possibly related ?
```
2) OCA\Files_Sharing\Tests\Controller\ShareAPIControllerTest::testUpdateLinkShareEnablePublicUpload with data set #0 (null, 'true', null, 'password')
Undefined variable $shareType
```

PVince81 commented:
@StCyr we'll take it from here, thanks a lot!

PVince81 commented on apps/sharebymail/tests/ShareByMailProviderTest.php:
```diff
@@ -267,24 +274,9 @@ public function testCreateSendPasswordByMailWithPasswordAndWithoutEnforcedPasswo
-			'initiatorEmail' => null,
-			'shareWith' => 'receiver@example.com',
-		]);
-		$this->mailer->expects($this->once())->method('send');
```
is this obsolete or more difficult to test ?

PVince81 replied:
probably we should duplicate the test to test the old behavior when the setting is set

PVince81 commented:
fixed the issue, rebased and squashed

next up: check the last issues, I hope at least CI passes now

PVince81 commented:
ran php:cs, rebuilt assets, merged suggestions from @CarlSchwan 

now hoping for green CI

PVince81 commented:
not that easy
```
"PHP Fatal error:  Declaration of OC\Share20\Share::setPasswordExpirationTime($passwordExpirationTime = null) must be compatible with OCP\Share\IShare::setPasswordExpirationTime($passwordExpirationTime = null): OCP\Share\IShare in /home/runner/work/server/server/lib/private/Share20/Share.php on line 469"
```

I'll have a look later

CarlSchwan commented on lib/public/Share/IShare.php:
```diff
@@ -448,6 +448,17 @@ public function setPassword($password);
+	 *
+	 * @return self The modified object
+	 */
+	public function setPasswordExpirationTime($passwordExpirationTime = null): self;
```
```suggestion
	public function setPasswordExpirationTime($passwordExpirationTime = null): IShare;
```

CarlSchwan commented on lib/public/Share/IShare.php:
```diff
@@ -448,6 +448,17 @@ public function setPassword($password);
+	/**
+	 * Get the password's expiration time of this share.
+	 */
+	public function getPasswordExpirationTime(): string;
```
```suggestion
	public function getPasswordExpirationTime(): ?string;
```

CarlSchwan commented on lib/public/Share/IShare.php:
```diff
@@ -448,6 +448,17 @@ public function setPassword($password);
+	/**
+	 * Set the password's expiration time of this share.
+	 *
+	 * @return self The modified object
```
```suggestion
	 * @return self The modified object
	 * @since 24.0.0
```

CarlSchwan commented on lib/public/Share/IShare.php:
```diff
@@ -448,6 +448,17 @@ public function setPassword($password);
+	public function setPasswordExpirationTime($passwordExpirationTime = null): self;
+
+	/**
+	 * Get the password's expiration time of this share.
```
```suggestion
	 * Get the password's expiration time of this share.
	 * @since 24.0.0
```

CarlSchwan commented on lib/public/AppFramework/AuthPublicShareController.php:
```diff
@@ -84,6 +84,29 @@ protected function showAuthFailed(): TemplateResponse {
+	/**
+	 * Generates a password
+	 *
+	 * @since 14.0.0
```
```suggestion
	 * @since 24.0.0
```

Should this event be part of the public API?

StCyr commented:
Hi @PVince81 @CarlSchwan,

I've fixed the unit tests (and added a new one to validate that password is not sent via email when temporary passwords are enabled).

Also, the test `testCreateSendPasswordByTalkWithEnforcedPasswordProtection` was failing because of an unexpected functional change imho. So, I've updated the function `ShareByMailProvider::create()` to allow sending password to the share owner in that case.

PVince81 commented:
@StCyr thanks a lot!

StCyr commented on apps/files_sharing/lib/Controller/ShareAPIController.php:
```diff
@@ -1513,6 +1521,32 @@ private function parseDate(string $expireDate): \DateTime {
+			}
+			$expirationTime = $now->add(new \DateInterval($expirationInterval));
+		} catch (\Exception $e) {
+			// Catches invalid format for system value 'share_temporary_password_expiration_interval'
```
I have added a logging and squashed my commits

CarlSchwan replied:
Ok cool, i will merge this tomorrow morning unless Vincent bewt me to it 😉

PVince81 reviewed (approved):
:+1:

PVince81 commented:
failing test unrelated and has just been fixed on master

merging

PVince81 commented:
seems a bug has sneaked in, the expiration date column never changes for me on master, raised here: https://github.com/nextcloud/server/issues/31951

PVince81 commented:
and some UX feedback: https://github.com/nextcloud/server/issues/31952
